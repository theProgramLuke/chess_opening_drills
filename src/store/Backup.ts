import _ from "lodash";
import fs from "graceful-fs";
import path from "path";

export const AgeSeparator = "-";

export class Backup {
  filePath: string;
  age: number;

  constructor(filePath: string) {
    this.filePath = filePath;
    this.age = _.parseInt(this.filePath.split(AgeSeparator)[1]);
  }

  async delete(): Promise<void> {
    await fs.unlink(this.filePath, _.noop);
  }

  async save(content: string): Promise<void> {
    await fs.mkdir(
      path.dirname(this.filePath),
      { recursive: true },
      async () => {
        await fs.writeFile(this.filePath, content, _.noop);
      }
    );
  }
}
