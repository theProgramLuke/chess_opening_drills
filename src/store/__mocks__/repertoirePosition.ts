import _ from "lodash";

import { Move } from "../move";
import { Side } from "../side";
import { TrainingEvent } from "../TrainingEvent";
import { TrainingMode } from "../trainingMode";
import { Turn } from "../turn";
import { FEN } from "chessground/types";
import { SavedRepertoirePosition } from "../repertoirePosition";

export class RepertoirePosition {
  fen: string;
  comment: string;
  parents: RepertoirePosition[];
  children: Move[];
  myTurn: boolean;
  forSide: Side;
  trainingHistory: TrainingEvent[];
  nextRepetitionTimestamp?: number | undefined;
  previousIntervalDays: number;
  intervalIndex: number;
  easinessFactor: number;
  SideToMove: () => Side;
  AddChild: (move: Move) => void;
  Unlink: () => void;
  IncludeForTrainingMode: (
    mode: TrainingMode,
    difficultyModeLimit?: number
  ) => boolean;
  AddTrainingEvent: (event: TrainingEvent) => void;
  VisitChildren: (visit: (child: RepertoirePosition) => void) => void;
  IsChildPosition: (maybeChild: RepertoirePosition) => boolean;
  RootPaths: () => Move[][];
  GetTurnLists: () => Turn[][];
  AsPgn: () => string;
  AsPgnMoveText: (turnCount?: number) => string;
  AsSaved: (childParentSource: RepertoirePosition[]) => SavedRepertoirePosition;

  constructor(fen: FEN, comment: string, forSide: Side) {
    this.fen = fen;
    this.comment = comment;
    this.forSide = forSide;
    this.parents = [];
    this.children = [];
    this.myTurn = false;
    this.forSide = Side.White;
    this.trainingHistory = [];
    this.nextRepetitionTimestamp = 0;
    this.previousIntervalDays = 0;
    this.intervalIndex = 0;
    this.easinessFactor = 0;
    this.SideToMove = jest.fn();
    this.AddChild = jest.fn();
    this.Unlink = jest.fn();
    this.IncludeForTrainingMode = jest.fn();
    this.AddTrainingEvent = jest.fn();
    this.VisitChildren = jest.fn();
    this.IsChildPosition = jest.fn();
    this.RootPaths = jest.fn();
    this.GetTurnLists = jest.fn();
    this.AsPgn = jest.fn();
    this.AsPgnMoveText = jest.fn();
    this.AsSaved = jest.fn();
  }
}
