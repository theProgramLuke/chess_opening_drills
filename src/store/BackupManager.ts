import fs from "graceful-fs";
import path from "path";
import _ from "lodash";

import { Backup } from "@/store/Backup";

export const day = 24 * 60 * 60 * 1000;
export const month = day * 31;
export const year = day * 365;

const lastBackupAge = (backups: Backup[]): number => {
  return _.min(_.map(backups, backup => backup.age)) || 0;
};

const trimBackups = (backups: Backup[], limit: number): Backup[] => {
  const sorted = _.sortBy(backups, backup => 0 - backup.age);

  while (sorted.length > limit) {
    const toDelete = sorted.pop();
    if (toDelete) {
      toDelete.delete();
    }
  }

  return sorted;
};

export class BackupManager {
  backupFolder: string;
  dailyBackups: Backup[];
  monthlyBackups: Backup[];
  yearlyBackups: Backup[];
  dailyLimit: number;
  monthlyLimit: number;
  yearlyLimit: number;
  createBackup: (filePath: string) => Backup;

  constructor(
    backupFolder: string,
    dailyLimit: number,
    monthlyLimit: number,
    yearlyLimit: number,
    listDirectory = (directory: string): string[] => fs.readdirSync(directory),
    createBackup = (filePath: string) => new Backup(filePath)
  ) {
    this.backupFolder = backupFolder;
    this.createBackup = createBackup;
    this.dailyBackups = this.discoverBackups(
      "daily",
      listDirectory,
      createBackup
    );
    this.monthlyBackups = this.discoverBackups(
      "monthly",
      listDirectory,
      createBackup
    );
    this.yearlyBackups = this.discoverBackups(
      "yearly",
      listDirectory,
      createBackup
    );
    this.dailyLimit = dailyLimit;
    this.monthlyLimit = monthlyLimit;
    this.yearlyLimit = yearlyLimit;
  }

  private discoverBackups(
    frequency: string,
    listDirectory: (directory: string) => string[],
    createBackup: (filePath: string) => Backup
  ): Backup[] {
    try {
      const files = listDirectory(path.join(this.backupFolder, frequency));
      return _.map(files, createBackup);
    } catch {
      return [];
    }
  }

  SaveBackup(content: string, now = _.now): void {
    const fileName = `settings-${now()}.json`;
    let backup: Backup;

    if (this.dailyLimit && now() - lastBackupAge(this.dailyBackups) >= day) {
      backup = this.createBackup(
        path.join(this.backupFolder, "daily", fileName)
      );
      this.dailyBackups.push(backup);
      backup.save(content);
    }
    this.dailyBackups = trimBackups(this.dailyBackups, this.dailyLimit);

    if (
      this.monthlyLimit &&
      now() - lastBackupAge(this.monthlyBackups) >= month
    ) {
      backup = this.createBackup(
        path.join(this.backupFolder, "monthly", fileName)
      );
      this.monthlyBackups.push(backup);
      backup.save(content);
    }
    this.monthlyBackups = trimBackups(this.monthlyBackups, this.monthlyLimit);

    if (this.yearlyLimit && now() - lastBackupAge(this.yearlyBackups) >= year) {
      backup = this.createBackup(
        path.join(this.backupFolder, "yearly", fileName)
      );
      this.yearlyBackups.push(backup);
      backup.save(content);
    }
    this.yearlyBackups = trimBackups(this.yearlyBackups, this.yearlyLimit);
  }
}
