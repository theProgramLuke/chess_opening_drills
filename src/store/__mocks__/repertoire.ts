import _ from "lodash";

import {
  RepertoirePosition,
  SavedRepertoirePosition
} from "@/store/repertoirePosition";
import { RepertoireTag, SavedRepertoireTag } from "@/store/repertoireTag";
import { Move } from "@/store/move";
import { PgnGame } from "pgn-parser";

export class SavedRepertoire {
  positions: SavedRepertoirePosition[];
  tags: SavedRepertoireTag[];

  constructor(
    positions: SavedRepertoirePosition[],
    tags: SavedRepertoireTag[]
  ) {
    this.positions = positions;
    this.tags = tags;
  }
}

export class Repertoire {
  positions: RepertoirePosition[];
  tags: RepertoireTag[];

  constructor(positions: RepertoirePosition[], tags: RepertoireTag[]) {
    this.positions = positions;
    this.tags = tags;
    this.AddMove = jest.fn();
    this.RemoveRepertoireTag = jest.fn();
    this.RemoveMove = jest.fn();
    this.AddFromGame = jest.fn();
    this.AsSaved = jest.fn();
  }

  AddMove: (parent: RepertoirePosition, move: Move) => void;

  RemoveRepertoireTag: (tagToRemove: RepertoireTag) => void;

  RemoveMove: (move: Move) => void;

  AddFromGame: (game: PgnGame) => void;

  AsSaved: () => SavedRepertoire;

  static FromSaved: (saved: SavedRepertoire) => Repertoire = jest.fn();
}
