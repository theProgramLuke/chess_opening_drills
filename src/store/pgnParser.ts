import { parse } from "@/store/pgnGrammar";

export interface PgnMove {
  moveNumber?: number;
  move?: PgnMove;
  nags?: string[];
  ravs?: string[];
  comments?: string;
}

export interface PgnHeader {
  name: string;
  value: string;
}

export interface PgnRav {
  moves: PgnMove[];
  result: string;
}

export interface PgnGame {
  commentsAboveHeader: string;
  headers: PgnHeader[];
  comments: string;
  moves: PgnMove[];
  result: string;
}

export function parsePgn(pgnText: string): PgnGame | string {
  try {
    return parse(pgnText) as PgnGame;
  } catch (error) {
    return error.message;
  }
}
