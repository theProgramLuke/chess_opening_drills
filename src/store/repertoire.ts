import _ from "lodash";
import { FEN } from "chessground/types";

import {
  RepertoirePosition,
  SavedRepertoirePosition
} from "@/store/repertoirePosition";
import { RepertoireTag, SavedRepertoireTag } from "@/store/repertoireTag";
import { Move } from "@/store/move";

export class SavedRepertoire {
  positions: SavedRepertoirePosition[];
  tags: SavedRepertoireTag[];

  constructor(
    positions: SavedRepertoirePosition[],
    tags: SavedRepertoireTag[]
  ) {
    this.positions = positions;
    this.tags = tags;
  }
}

export class Repertoire {
  positions: RepertoirePosition[];
  tags: RepertoireTag[];

  constructor(positions: RepertoirePosition[], tags: RepertoireTag[]) {
    this.positions = positions;
    this.tags = tags;
  }

  AddMove(parent: RepertoirePosition, move: Move): void {
    const existingPosition = this.LookupRepertoirePosition(move.position.fen);
    if (existingPosition) {
      move.position = existingPosition;
    }

    parent.AddChild(move);

    if (!existingPosition) {
      this.positions.push(move.position);
    }
  }

  private LookupRepertoirePosition(fen: FEN): RepertoirePosition | undefined {
    return _.find(this.positions, (position: RepertoirePosition) => {
      return fen === position.fen;
    });
  }

  AsSaved(): SavedRepertoire {
    const savedPositions = _.map(this.positions, position =>
      position.AsSaved(this.positions)
    );

    const loadedTags = _.map(this.tags, tag => tag.AsSaved(this.positions));

    return new SavedRepertoire(savedPositions, loadedTags);
  }

  static FromSaved(saved: SavedRepertoire): Repertoire {
    const loadedPositions = _.map(saved.positions, savedPosition =>
      RepertoirePosition.FromSaved(savedPosition)
    );

    _.forEach(saved.positions, (savedPosition, index) =>
      _.forEach(savedPosition.children, savedMove =>
        loadedPositions[index].AddChild(
          new Move(savedMove.san, loadedPositions[savedMove.positionId])
        )
      )
    );

    const loadedTags = _.map(saved.tags, savedTag =>
      RepertoireTag.FromSaved(savedTag, loadedPositions)
    );

    return new Repertoire(loadedPositions, loadedTags);
  }
}
