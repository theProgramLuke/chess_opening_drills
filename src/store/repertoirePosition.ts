import _ from "lodash";
import { FEN } from "chessground/types";
import { Chess } from "chess.js";

import { Move, SavedMove } from "@/store/move";
import { Side } from "@/store/side";
import { Turn } from "@/store/turn";

export class SavedRepertoirePosition {
  fen: FEN;
  comment: string;
  parentIds: number[];
  children: SavedMove[];
  myTurn: boolean;
  forSide: Side;

  constructor(
    fen: FEN,
    comment: string,
    parentIds: number[],
    children: SavedMove[],
    myTurn: boolean,
    forSide: Side
  ) {
    this.fen = fen;
    this.comment = comment;
    this.parentIds = parentIds;
    this.children = children;
    this.myTurn = myTurn;
    this.forSide = forSide;
  }
}

export class RepertoirePosition {
  fen: FEN;
  comment: string;
  parents: RepertoirePosition[];
  children: Move[];
  myTurn: boolean;
  forSide: Side;

  constructor(fen: FEN, myTurn: boolean, comment: string, forSide: Side) {
    this.fen = fen;
    this.comment = comment;
    this.myTurn = myTurn;
    this.forSide = forSide;
    this.parents = [];
    this.children = [];
  }

  SideToMove(): Side {
    return Chess(this.fen).turn() === "w" ? Side.White : Side.Black;
  }

  AddChild(move: Move): void {
    const samePosition = this.fen == move.position.fen;
    const alreadyChild = !_.isEmpty(
      _.find(
        this.children,
        (child: Move) => move.position.fen === child.position.fen
      )
    );
    if (!samePosition && !alreadyChild) {
      this.children.push(move);
      move.position.parents.push(this);
    }
  }

  private RootPaths(): Move[][] {
    const collector: Move[][] = [];

    if (!_.isEmpty(this.parents)) {
      this.RootPathsRecursive([], collector);
    }

    return collector;
  }

  private RootPathsRecursive(path: Move[], collector: Move[][]): void {
    if (_.isEmpty(this.parents)) {
      collector.push(path);
    } else {
      _.forEach(this.parents, parent => {
        const branch = _.clone(path);
        const parentMove = parent.getChildMove(this);

        if (parentMove) {
          branch.unshift(parentMove);
        }

        parent.RootPathsRecursive(branch, collector);
      });
    }
  }

  private getChildMove(other: RepertoirePosition): Move | undefined {
    return _.find(this.children, child => child.position === other);
  }

  GetTurnLists(): Array<Array<Turn>> {
    const moveLists: Array<Array<Turn>> = [];

    const paths = this.RootPaths();

    _.forEach(paths, (path: Array<Move>) => {
      const turns: Array<Turn> = [];

      _.forEach(_.range(0, path.length, 2), (i: number) => {
        const whiteMove = path[i];

        if (path.length === i) {
          turns.push(new Turn(whiteMove));
        } else {
          const blackMove = path[i + 1];
          turns.push(new Turn(whiteMove, blackMove));
        }
      });

      moveLists.push(turns);
    });

    return moveLists;
  }

  AsPgnMoveText(turnCount = 1): string {
    if (_.isEmpty(this.children)) {
      return "";
    }

    const side = this.SideToMove();
    let pgnMoveText = "";
    let childTurnCount = turnCount;

    if (side === Side.White) {
      // White made this move.
      pgnMoveText += turnCount + ". ";
      childTurnCount++;
    }

    const mainLine = _.first(this.children);
    if (mainLine) {
      pgnMoveText += mainLine.san;
    }

    const variations = _.tail(this.children);

    _.forEach(variations, child => {
      pgnMoveText += " ( "; // start variation
      pgnMoveText += turnCount;
      pgnMoveText += side === Side.White ? ". " : "... ";
      pgnMoveText += child.san;
      pgnMoveText += " ";
      pgnMoveText += child.position.AsPgnMoveText(childTurnCount);
      pgnMoveText += " )"; // end variation
    });

    if (mainLine) {
      pgnMoveText += " ";
      pgnMoveText += mainLine.position.AsPgnMoveText(childTurnCount);
    }

    return _.trim(pgnMoveText);
  }

  AsSaved(childParentSource: RepertoirePosition[]): SavedRepertoirePosition {
    const savedChildren = _.map(
      this.children,
      child =>
        new SavedMove(child.san, _.indexOf(childParentSource, child.position))
    );

    const parentIds = _.map(this.parents, parent =>
      _.indexOf(childParentSource, parent)
    );

    return new SavedRepertoirePosition(
      this.fen,
      this.comment,
      parentIds,
      savedChildren,
      this.myTurn,
      this.forSide
    );
  }

  static FromSaved(saved: SavedRepertoirePosition): RepertoirePosition {
    return new RepertoirePosition(
      saved.fen,
      saved.myTurn,
      saved.comment,
      saved.forSide
    );
  }
}
