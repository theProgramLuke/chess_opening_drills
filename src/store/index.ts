import Vue from "vue";
import Vuex from "vuex";

import { RepertoirePosition } from "@/store/repertoirePosition";
import { Move } from "./move";
import { GetPersistantStorage } from "./storage";
import { Repertoire } from "./repertoire";
import { RepertoireTag } from "./repertoireTag";
import { Side } from "./side";
import { TrainingEvent } from "./TrainingEvent";
import { PgnGame } from "./pgnParser";

Vue.use(Vuex);

const storage = GetPersistantStorage();

export default new Vuex.Store({
  state: {
    darkMode: storage.get("darkMode"),
    primary: storage.get("primary"),
    secondary: storage.get("secondary"),
    accent: storage.get("accent"),
    error: storage.get("error"),
    warning: storage.get("warning"),
    info: storage.get("info"),
    success: storage.get("success"),
    boardTheme: storage.get("boardTheme"),
    pieceTheme: storage.get("pieceTheme"),
    whiteRepertoire: Repertoire.FromSaved(storage.get("whiteRepertoire")),
    blackRepertoire: Repertoire.FromSaved(storage.get("blackRepertoire"))
  },

  mutations: {
    setDarkMode: (state, darkMode): void => {
      state.darkMode = darkMode;
      storage.set("darkMode", darkMode);
    },

    setBoardTheme: (state, boardTheme): void => {
      state.boardTheme = boardTheme;
      storage.set("boardTheme", boardTheme);
    },

    setPieceTheme: (state, pieceTheme): void => {
      state.pieceTheme = pieceTheme;
      storage.set("pieceTheme", pieceTheme);
    },

    setColor: function(
      state,
      payload: {
        colorToSet:
          | "primary"
          | "secondary"
          | "accent"
          | "error"
          | "warning"
          | "info"
          | "success";
        color: string;
      }
    ): void {
      state[payload.colorToSet] = payload.color;
      storage.set(payload.colorToSet, payload.color);
    },

    addRepertoirePosition: (
      state,
      payload: { parent: RepertoirePosition; newMove: Move }
    ): void => {
      if (payload.parent && payload.newMove) {
        const repertoire =
          payload.parent.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.parent.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        repertoire.AddMove(payload.parent, payload.newMove);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    addRepertoireTag: (
      state,
      payload: { parent: RepertoireTag; tag: RepertoireTag }
    ): void => {
      if (payload.parent && payload.tag) {
        const repertoire =
          payload.parent.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.parent.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        payload.parent.AddChild(payload.tag);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    addTrainingEvent: (
      state,
      payload: { position: RepertoirePosition; event: TrainingEvent }
    ): void => {
      if (payload.position && payload.event) {
        const repertoire =
          payload.position.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.position.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        payload.position.AddTrainingEvent(payload.event);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    removeRepertoireTag: (state, tag: RepertoireTag): void => {
      const repertoire =
        tag.forSide === Side.White
          ? state.whiteRepertoire
          : state.blackRepertoire;
      const repertoireKey =
        tag.forSide === Side.White ? "whiteRepertoire" : "blackRepertoire";

      repertoire.RemoveRepertoireTag(tag);

      storage.set(repertoireKey, repertoire.AsSaved());
    },

    removeRepertoireMove: (state, move: Move): void => {
      const repertoire =
        move.position.forSide === Side.White
          ? state.whiteRepertoire
          : state.blackRepertoire;
      const repertoireKey =
        move.position.forSide === Side.White
          ? "whiteRepertoire"
          : "blackRepertoire";

      repertoire.RemoveMove(move);

      storage.set(repertoireKey, repertoire.AsSaved());
    },

    addPositionsFromGame(
      state,
      payload: { forSide: Side; game: PgnGame }
    ): void {
      if (payload.game) {
        const repertoire =
          payload.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        repertoire.AddFromGame(payload.game);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    clearStorage(): void {
      storage.clear();
    }
  },

  actions: {},

  modules: {},

  strict: process.env.NODE_ENV !== "production"
});
