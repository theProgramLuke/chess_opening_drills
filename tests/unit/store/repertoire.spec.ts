import _ from "lodash";

import { Repertoire } from "@/store/repertoire";
import { RepertoirePosition } from "@/store/repertoirePosition";
import { parsePgn } from "@/store/pgnParser";
import {
  ResetTestRepertoire,
  repertoire,
  start,
  e3,
  e3e6,
  e3e6d3,
  d3,
  d3e6,
  LinkTestPositions,
  ExpectedSavedRepertoire,
  VariationC,
  VariationA
} from "./testDataRepertoire";

beforeEach(() => {
  ResetTestRepertoire();
});

describe("Repertoire", () => {
  describe("AsSaved", () => {
    it("transforms to a saved repertoire", () => {
      LinkTestPositions();

      const saved = repertoire.AsSaved();

      expect(saved).toMatchSnapshot();
    });
  });

  describe("FromSaved", () => {
    it("generates a real repertoire from a saved one", () => {
      LinkTestPositions();

      const loaded = Repertoire.FromSaved(repertoire.AsSaved());

      expect(loaded).toEqual(repertoire);
      expect(loaded.positions[0].GetTurnLists).toBeDefined();
    });
  });

  describe("AddMoves", () => {
    it("adds position given position not in repertoire positions", () => {
      repertoire.AddMove(start, e3);

      expect(repertoire.positions).toEqual([start, e3.position]);
    });

    it("adds the move as a child of the parent", () => {
      const repertoire = new Repertoire([start], []);

      repertoire.AddMove(start, e3);

      expect(start.children).toEqual([e3]);
    });

    it("does not add duplicate position given a transposition", () => {
      LinkTestPositions();

      _.forEach(
        [
          start,
          e3.position,
          e3e6.position,
          e3e6d3.position,
          d3.position,
          d3e6.position
        ],
        (position: RepertoirePosition) =>
          expect(repertoire.positions).toContain(position)
      );
    });
  });

  describe("RemoveMove", () => {
    beforeEach(LinkTestPositions);

    it("deletes all orphaned child positions", () => {
      repertoire.RemoveMove(d3);

      expect(start.children.length).toBe(1);
      expect(repertoire.positions.length).toBe(6);
    });
  });

  describe("RemoveRepertoireTag", () => {
    it("should unlink a repertoire tag", () => {
      LinkTestPositions();

      repertoire.RemoveRepertoireTag(VariationC);

      expect(repertoire.tags[0].children).toEqual([VariationA]);
    });
  });

  describe("AddFromGame", () => {
    it("adds all game variations", () => {
      const parsedGame = parsePgn(
        '[White "me"]\n[Black "you"]\n1. e4 e5 (1. ...c5) *'
      )[0];
      const repertoire = new Repertoire([start], []);

      repertoire.AddFromGame(parsedGame);

      expect(_.map(start.children, child => child.san)).toEqual(["e4"]);
      expect(
        _.map(start.children[0].position.children, child => child.san)
      ).toEqual(["c5", "e5"]);
    });

    it("added positions can be removed", () => {
      const parsedGame = parsePgn(
        '[White "me"]\n[Black "you"]\n1. e4 e5 2. Nf3 Nc6 3. Bc4 Bc5 (3... Nf6) *'
      )[0];
      const repertoire = new Repertoire([start], []);

      repertoire.AddFromGame(parsedGame);
      repertoire.RemoveMove(start.children[0]);

      expect(repertoire.positions.length).toEqual(1);
    });
  });
});
