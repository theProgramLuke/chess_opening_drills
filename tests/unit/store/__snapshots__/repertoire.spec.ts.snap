// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repertoire AsSaved transforms to a saved repertoire 1`] = `
SavedRepertoire {
  "positions": Array [
    SavedRepertoirePosition {
      "children": Array [
        SavedMove {
          "positionId": 1,
          "san": "e3",
        },
        SavedMove {
          "positionId": 6,
          "san": "d3",
        },
      ],
      "comment": "",
      "easinessFactor": 2.5,
      "fen": "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
      "forSide": 0,
      "intervalIndex": 0,
      "myTurn": true,
      "nextRepetitionTimestamp": undefined,
      "parentIds": Array [],
      "previousIntervalDays": 0,
      "trainingHistory": Array [],
    },
    SavedRepertoirePosition {
      "children": Array [
        SavedMove {
          "positionId": 2,
          "san": "e6",
        },
      ],
      "comment": "",
      "easinessFactor": 2.5,
      "fen": "rnbqkbnr/pppppppp/8/8/8/4P3/PPPP1PPP/RNBQKBNR b KQkq - 0 1",
      "forSide": 0,
      "intervalIndex": 0,
      "myTurn": false,
      "nextRepetitionTimestamp": undefined,
      "parentIds": Array [
        0,
      ],
      "previousIntervalDays": 0,
      "trainingHistory": Array [],
    },
    SavedRepertoirePosition {
      "children": Array [
        SavedMove {
          "positionId": 3,
          "san": "d3",
        },
        SavedMove {
          "positionId": 4,
          "san": "e4",
        },
      ],
      "comment": "",
      "easinessFactor": 2.5,
      "fen": "rnbqkbnr/pppp1ppp/4p3/8/8/4P3/PPPP1PPP/RNBQKBNR w KQkq - 0 2",
      "forSide": 0,
      "intervalIndex": 0,
      "myTurn": true,
      "nextRepetitionTimestamp": undefined,
      "parentIds": Array [
        1,
      ],
      "previousIntervalDays": 0,
      "trainingHistory": Array [],
    },
    SavedRepertoirePosition {
      "children": Array [],
      "comment": "",
      "easinessFactor": 2.5,
      "fen": "rnbqkbnr/pppp1ppp/4p3/8/8/3PP3/PPP2PPP/RNBQKBNR b KQkq - 0 2",
      "forSide": 0,
      "intervalIndex": 0,
      "myTurn": false,
      "nextRepetitionTimestamp": undefined,
      "parentIds": Array [
        2,
        7,
      ],
      "previousIntervalDays": 0,
      "trainingHistory": Array [],
    },
    SavedRepertoirePosition {
      "children": Array [
        SavedMove {
          "positionId": 5,
          "san": "e5",
        },
      ],
      "comment": "",
      "easinessFactor": 2.5,
      "fen": "rnbqkbnr/pppp1ppp/4p3/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq - 0 2",
      "forSide": 0,
      "intervalIndex": 0,
      "myTurn": false,
      "nextRepetitionTimestamp": undefined,
      "parentIds": Array [
        2,
      ],
      "previousIntervalDays": 0,
      "trainingHistory": Array [],
    },
    SavedRepertoirePosition {
      "children": Array [],
      "comment": "",
      "easinessFactor": 2.5,
      "fen": "rnbqkbnr/pppp1ppp/8/4p3/4P3/8/PPPP1PPP/RNBQKBNR w KQkq - 0 3",
      "forSide": 0,
      "intervalIndex": 0,
      "myTurn": true,
      "nextRepetitionTimestamp": undefined,
      "parentIds": Array [
        4,
      ],
      "previousIntervalDays": 0,
      "trainingHistory": Array [],
    },
    SavedRepertoirePosition {
      "children": Array [
        SavedMove {
          "positionId": 7,
          "san": "e6",
        },
      ],
      "comment": "",
      "easinessFactor": 2.5,
      "fen": "rnbqkbnr/pppppppp/8/8/8/3P4/PPP1PPPP/RNBQKBNR b KQkq - 0 1",
      "forSide": 0,
      "intervalIndex": 0,
      "myTurn": false,
      "nextRepetitionTimestamp": undefined,
      "parentIds": Array [
        0,
      ],
      "previousIntervalDays": 0,
      "trainingHistory": Array [],
    },
    SavedRepertoirePosition {
      "children": Array [
        SavedMove {
          "positionId": 3,
          "san": "e3",
        },
      ],
      "comment": "",
      "easinessFactor": 2.5,
      "fen": "rnbqkbnr/pppp1ppp/4p3/8/8/3P4/PPP1PPPP/RNBQKBNR w KQkq - 0 2",
      "forSide": 0,
      "intervalIndex": 0,
      "myTurn": true,
      "nextRepetitionTimestamp": undefined,
      "parentIds": Array [
        6,
      ],
      "previousIntervalDays": 0,
      "trainingHistory": Array [],
    },
  ],
  "tags": Array [
    SavedRepertoireTag {
      "children": Array [
        SavedRepertoireTag {
          "children": Array [
            SavedRepertoireTag {
              "children": Array [],
              "fen": "rnbqkbnr/pppp1ppp/4p3/8/8/4P3/PPPP1PPP/RNBQKBNR w KQkq - 0 2",
              "forSide": 0,
              "id": "3",
              "name": "Variation AB",
              "positionId": 2,
            },
          ],
          "fen": "rnbqkbnr/pppp1ppp/4p3/8/8/4P3/PPPP1PPP/RNBQKBNR w KQkq - 0 2",
          "forSide": 0,
          "id": "2",
          "name": "Variation A",
          "positionId": 2,
        },
        SavedRepertoireTag {
          "children": Array [],
          "fen": "rnbqkbnr/pppp1ppp/4p3/8/8/4P3/PPPP1PPP/RNBQKBNR w KQkq - 0 2",
          "forSide": 0,
          "id": "4",
          "name": "Variation C",
          "positionId": 2,
        },
      ],
      "fen": "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
      "forSide": 0,
      "id": "0",
      "name": "White",
      "positionId": 0,
    },
    SavedRepertoireTag {
      "children": Array [],
      "fen": "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
      "forSide": 1,
      "id": "1",
      "name": "Black",
      "positionId": 0,
    },
  ],
}
`;
